pkg/datasource/data.go

```diff
@@ -318,26 +403,61 @@ func (ds *dataSource) Subscribe(fn DataFunc, priority int) {
 	ds.lock.Lock()
 	defer ds.lock.Unlock()
 
-	ds.subscriptions = append(ds.subscriptions, &subscription{
+	ds.subscriptionsData = append(ds.subscriptionsData, &subscriptionData{
 		priority: priority,
 		fn:       fn,
 	})
-	sort.SliceStable(ds.subscriptions, func(i, j int) bool {
-		return ds.subscriptions[i].priority < ds.subscriptions[j].priority
+	sort.SliceStable(ds.subscriptionsData, func(i, j int) bool {
+		return ds.subscriptionsData[i].priority < ds.subscriptionsData[j].priority
 	})
 }
 
-func (ds *dataSource) EmitAndRelease(d Data) error {
-	for _, sub := range ds.subscriptions {
-		err := sub.fn(ds, d)
+func (ds *dataSource) SubscribePacket(fn PacketFunc, priority int) {
+	if fn == nil {
+		return
+	}
+
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	ds.subscriptionsPacket = append(ds.subscriptionsPacket, &subscriptionPacket{
+		priority: priority,
+		fn:       fn,
+	})
+	sort.SliceStable(ds.subscriptionsPacket, func(i, j int) bool {
+		return ds.subscriptionsPacket[i].priority < ds.subscriptionsPacket[j].priority
+	})
+}
+
+func (ds *dataSource) EmitAndRelease(p Packet) error {
+	defer ds.Release(p)
+	for _, sub := range ds.subscriptionsPacket {
+		err := sub.fn(ds, p)
 		if err != nil {
 			return err
 		}
 	}
+	for _, sub := range ds.subscriptionsData {
+		switch ds.dType {
+		case TypeSingle:
+			err := sub.fn(ds, p.(Data))
+			if err != nil {
+				return err
+			}
+		case TypeArray:
+			iter := p.(PacketArray).Iterate()
+			for iter.Next() {
+				err := sub.fn(ds, iter.Data)
+				if err != nil {
+					return err
+				}
+			}
```

Comment:
I'd prefer not to use `Iterate()` here to avoid the allocation in that function - we can iterate directly over the array, right? (I think in general that direct iteration using a `.Count()` and then `.GetElement(idx)` could be more efficient (avoiding the allocation as well), but I don't have concrete numbers on that.

Reply:
Here in particular, I agree as it's within the datasource package, so the underlay struct (dataArray) is known.

Reply:
BTW, I applied the same for `Dump`.